experimental/token/cursor.go

```diff
@@ -95,46 +90,98 @@ func (c *Cursor) Rewind(mark CursorMark) {
 	if c != mark.owner {
 		panic("protocompile/ast: rewound cursor using the wrong cursor's mark")
 	}
-	c.start = mark.start
 	c.idx = mark.idx
+	c.isBackwards = mark.isBackwards
 }
 
-// Peek returns the next token in the sequence, if there is one.
+// PeekSkippable returns the current token in the sequence, if there is one.
 // This may return a skippable token.
 //
 // Returns the zero token if this cursor is at the end of the stream.
 func (c *Cursor) PeekSkippable() Token {
 	if c == nil {
 		return Zero
 	}
-
 	if c.IsSynthetic() {
-		if c.idx == len(c.stream) {
+		if c.idx < 0 || c.idx >= len(c.stream) {
 			return Zero
 		}
 		return c.stream[c.idx].In(c.Context())
 	}
-	if c.start >= c.end {
+	if c.idx < 1 || c.idx > len(c.Context().Stream().nats) {
 		return Zero
 	}
-	return c.start.In(c.Context())
+	tok := ID(c.idx).In(c.Context())
+	if !c.isBackwards && tok.nat().IsClose() {
+		return Zero // Reached the end.
+	}
+	return tok
 }
 
-// Pop returns the next skippable token in the sequence, and advances the cursor.
-func (c *Cursor) PopSkippable() Token {
+// BeforeSkippable returns the token before the current token in the sequence, if there is one.
+// This may return a skippable token.
+//
+// Returns the zero token if this cursor is at the beginning of the stream.
+func (c *Cursor) BeforeSkippable() Token {
+	if c == nil {
+		return Zero
+	}
+	if c.IsSynthetic() {
+		if c.idx-1 < 0 || c.idx-1 >= len(c.stream) {
+			return Zero
+		}
+		return c.stream[c.idx-1].In(c.Context())
+	}
+	idx := c.idx - 1
+	if c.isBackwards && c.idx >= 1 && c.idx <= len(c.Context().Stream().nats) {
```

Comment:
Can you spill `c.Context().Stream()` into its own variable somewhere? It would help make this a little more readable.

Also, one way that you can simplify this is to use `slicesx.Get(s.nats, idx)`.

Reply:
Switched to slicesx.Get.